File: android/src/main/java/expo/modules/callkittelecom/events/CallEventEmitter.kt
package expo.modules.callkittelecom.events

import android.content.Context
import android.content.Intent
import expo.modules.callkittelecom.utils.CallKitTelecomLog
import java.time.Instant
import java.time.format.DateTimeFormatter
import org.json.JSONObject

private data class QueuedEvent(val body: Map<String, Any?>, val timestamp: Instant)

/**
 * Event bridge between Android native call state and the Expo JS module.
 *
 * Behavior:
 * - Tracks which individual events are being observed by JS
 * - Queues events that arrive before observers mount
 * - Flushes queued events with `{ meta: { flushed: true } }`
 * - Broadcasts events that can't reach JS at all (dropped, not queued) when a broadcast context is
 *   set, so a JS-less process can still react (see [setBroadcastContext])
 *
 * All mutable state is guarded by [lock] for thread safety.
 */
object CallEventEmitter {
    private const val TAG = "ExpoCallKitTelecom.Emitter"

    /**
     * Package-internal broadcast action for call events that couldn't reach a live JS observer.
     * Apps receive it with a manifest receiver, registered via the `androidEventReceiver`
     * config-plugin prop.
     */
    const val ACTION_CALL_EVENT = "expo.modules.callkittelecom.ACTION_CALL_EVENT"

    private val lock = Any()
    private val observingEvents = mutableSetOf<String>()
    private val eventQueues = mutableMapOf<String, MutableList<QueuedEvent>>()
    private val queueLimits = mutableMapOf<String, Int?>()

    var defaultQueueLimit: Int? = 0

    @Volatile private var sender: ((String, Map<String, Any?>) -> Unit)? = null

    @Volatile private var broadcastContext: Context? = null

    /** Sets or clears the active event sender provided by the Expo module. */
    fun setSender(eventSender: ((String, Map<String, Any?>) -> Unit)?) {
        sender = eventSender
    }

    /**
     * Enables (or disables, with `null`) the broadcast path by providing an application context.
     *
     * When set, any event that [send] *drops* — i.e. one that can't reach a live JS observer and
     * isn't queued for cold-start replay (queue limit 0) — is emitted as a package-internal
     * [ACTION_CALL_EVENT] broadcast, so a process started by a push or Telecom with no React
     * context can still react (e.g. notify a backend of a killed-app decline). Queued events are
     * not broadcast: they flush to JS once observed, so the queue already covers them.
     *
     * Broadcasting sits next to the queue — the emitter's existing handling for events JS can't
     * receive yet — rather than behind an injected hook, since the actual consumer (the app's
     * manifest receiver) is already decoupled by the OS and there is nothing to inject.
     */
    fun setBroadcastContext(context: Context?) {
        broadcastContext = context?.applicationContext
    }

    /**
     * Configures queue size for a specific event.
     *
     * `null` means unlimited queueing, `0` disables queueing.
     */
    fun setQueueLimit(eventName: String, limit: Int?) {
        synchronized(lock) { queueLimits[eventName] = limit }
    }

    /**
     * Sends an event to JS if it is currently observed, or queues it otherwise.
     *
     * All delivered events are augmented with a `meta` object containing timestamp and flush
     * status.
     *
     * @return true when the event was delivered to a live JS observer, false when it was queued (or
     *   dropped by the event's queue limit). When the event was *dropped* (it will never reach JS)
     *   and a broadcast context is set (see [setBroadcastContext]), it is also broadcast so native
     *   code can deliver out-of-band — e.g. the app process was started by a push or Telecom with
     *   no React context. Events that were queued are not broadcast: they flush to JS once
     *   observed, so broadcasting them too would double-deliver.
     */
    fun send(eventName: String, body: Map<String, Any?>): Boolean {
        val timestamp = Instant.now()
        val senderRef = sender
        val isObserving = synchronized(lock) { observingEvents.contains(eventName) }

        if (senderRef != null && isObserving) {
            CallKitTelecomLog.d(TAG) { "Sending event to JS - name: $eventName" }
            senderRef(eventName, buildEventBody(body, flushed = false, timestamp = timestamp))
            return true
        }
        val queued = queueEvent(eventName, body, timestamp)
        if (!queued) {
            broadcastContext?.let { context ->
                broadcastUndelivered(
                    context,
                    eventName,
                    buildEventBody(body, flushed = false, timestamp = timestamp),
                )
            }
        }
        return false
    }

    /**
     * Broadcasts an undelivered event so a JS-less process can still react.
     *
     * Fires a package-internal ([Intent.setPackage]) [ACTION_CALL_EVENT] broadcast carrying the
     * event name and the JS-shaped body (the same one JS would have received) as a JSON `payload`.
     * Receivers discriminate on the `eventName` extra. Failures are swallowed (warn-logged): a
     * dropped broadcast must never break event emission.
     */
    private fun broadcastUndelivered(context: Context, eventName: String, body: Map<String, Any?>) {
        try {
            val intent =
                Intent(ACTION_CALL_EVENT)
                    .setPackage(context.packageName)
                    .putExtra("eventName", eventName)
                    .putExtra("payload", JSONObject(body).toString())
            context.sendBroadcast(intent)
            CallKitTelecomLog.d(TAG) {
                "Broadcast undelivered event (no live JS observer) - name: $eventName"
            }
        } catch (e: Exception) {
            CallKitTelecomLog.w(TAG) { "Event broadcast failed - name: $eventName: ${e.message}" }
        }
    }

    /** Marks an event as observed and flushes any pending queue for that event. */
    fun startObserving(eventName: String) {
        val queueCount: Int
        synchronized(lock) {
            queueCount = eventQueues[eventName]?.size ?: 0
            observingEvents.add(eventName)
        }
        CallKitTelecomLog.d(TAG) {
            "Start observing - event: $eventName, queuedEvents: $queueCount"
        }
        flushQueue(eventName)
    }

    /** Marks an event as no longer observed by JS. */
    fun stopObserving(eventName: String) {
        CallKitTelecomLog.d(TAG) { "Stop observing - event: $eventName" }
        synchronized(lock) { observingEvents.remove(eventName) }
    }

    /** Adds native event metadata used by TypeScript event types. */
    private fun buildEventBody(
        body: Map<String, Any?>,
        flushed: Boolean,
        timestamp: Instant,
    ): Map<String, Any?> {
        val result = body.toMutableMap()
        result["meta"] =
            mapOf(
                "flushed" to flushed,
                "timestamp" to DateTimeFormatter.ISO_INSTANT.format(timestamp),
            )
        return result
    }

    /**
     * Queues an event and enforces per-event queue limits (drop oldest first).
     *
     * @return true when the event was queued (and will flush to JS once observed), false when
     *   queueing is disabled for it (limit 0) so it was dropped and will never reach JS.
     */
    private fun queueEvent(name: String, body: Map<String, Any?>, timestamp: Instant): Boolean =
        synchronized(lock) {
            val limit = queueLimits[name] ?: defaultQueueLimit
            if (limit == 0) {
                CallKitTelecomLog.d(TAG) { "Dropping event (queueing disabled) - name: $name" }
                return@synchronized false
            }

            val queue = eventQueues.getOrPut(name) { mutableListOf() }
            queue += QueuedEvent(body = body, timestamp = timestamp)

            if (limit != null && queue.size > limit) {
                val dropCount = queue.size - limit
                repeat(dropCount) { queue.removeAt(0) }
                CallKitTelecomLog.d(TAG) {
                    "Queueing event (dropped $dropCount old) - name: $name, queueSize: ${queue.size}"
                }
            } else {
                CallKitTelecomLog.d(TAG) {
                    "Queueing event (JS not listening) - name: $name, queueSize: ${queue.size}"
                }
            }
            true
        }

    /** Flushes all queued events for a single event name. */
    private fun flushQueue(eventName: String) {
        val senderRef = sender ?: return
        val queue = synchronized(lock) { eventQueues.remove(eventName) } ?: return
        if (queue.isEmpty()) return

        CallKitTelecomLog.d(TAG) {
            "Flushing event queue - event: $eventName, count: ${queue.size}"
        }
        queue.forEach { event ->
            senderRef(
                eventName,
                buildEventBody(event.body, flushed = true, timestamp = event.timestamp),
            )
        }
    }
}

File: ios/Managers/CallEventEmitter.swift
import ExpoModulesCore
import Foundation

/// A queued event waiting to be sent to JS.
private struct QueuedEvent {
  let body: [String: Any]
  let timestamp: Date
}

/// Manages event delivery to the Expo module, buffering events until JS is ready.
///
/// CallKit events can fire before the JS layer has mounted its listeners.
/// This class queues those events and flushes them once JS starts observing.
///
/// Supports per-event configuration:
/// - Track which events are being observed independently
/// - Configure queue limits per event (0 = no queueing, nil = unlimited)
@MainActor
final class CallEventEmitter {
  static let shared = CallEventEmitter()

  /// The Expo module to send events to. Set this when the module initializes.
  weak var module: ExpoCallKitTelecomModule?

  /// Events currently being observed by JS.
  private var observingEvents: Set<String> = []

  /// Per-event queues for buffering events before JS is ready.
  private var eventQueues: [String: [QueuedEvent]] = [:]

  /// Per-event queue limits. `nil` = unlimited, `0` = no queueing.
  private var queueLimits: [String: Int] = [:]

  /// Default queue limit for events without explicit configuration.
  /// Set to `nil` for unlimited, `0` to disable queueing by default.
  var defaultQueueLimit: Int? = 0

  private init() {}

  // MARK: - Configuration

  /// Set the queue limit for a specific event.
  /// - Parameters:
  ///   - eventName: The event name to configure
  ///   - limit: Maximum queued events. `nil` = unlimited, `0` = no queueing.
  func setQueueLimit(for eventName: String, limit: Int?) {
    if let limit = limit {
      queueLimits[eventName] = limit
    } else {
      queueLimits.removeValue(forKey: eventName)
    }
  }

  /// Set the queue limit for a specific event type (type-safe).
  /// - Parameters:
  ///   - eventType: The event type to configure
  ///   - limit: Maximum queued events. `nil` = unlimited, `0` = no queueing.
  func setQueueLimit<E: CallEvent>(for eventType: E.Type, limit: Int?) {
    setQueueLimit(for: E.name, limit: limit)
  }

  /// Set queue limits for multiple events at once.
  func setQueueLimits(_ limits: [String: Int?]) {
    for (eventName, limit) in limits {
      setQueueLimit(for: eventName, limit: limit)
    }
  }

  /// Disable queueing for events that should only be delivered in real-time.
  /// Useful for events like mute/hold changes that are only relevant when observed.
  func disableQueueing<E: CallEvent>(for eventType: E.Type) {
    setQueueLimit(for: E.name, limit: 0)
  }

  // MARK: - Status

  /// Check if a specific event is currently being observed.
  func isObserving(eventName: String) -> Bool {
    observingEvents.contains(eventName)
  }

  /// Check if a specific event type is currently being observed.
  func isObserving<E: CallEvent>(_ eventType: E.Type) -> Bool {
    isObserving(eventName: E.name)
  }

  /// Get the current queue count for a specific event.
  func queueCount(for eventName: String) -> Int {
    eventQueues[eventName]?.count ?? 0
  }

  /// Get the total count of all queued events.
  var totalQueueCount: Int {
    eventQueues.values.reduce(0) { $0 + $1.count }
  }

  // MARK: - Public API

  /// Send a type-safe event to JS, or buffer it if JS isn't listening yet.
  func send<E: CallEvent>(_ event: E) {
    let eventName = E.name
    let timestamp = Date()

    if observingEvents.contains(eventName), let module = module {
      Log.call.debug("Sending event to JS - name: \(eventName)")
      let body = buildEventBody(
        event.body,
        flushed: false,
        timestamp: timestamp
      )
      module.sendEvent(eventName, body)
    } else {
      queueEvent(name: eventName, body: event.body, timestamp: timestamp)
    }
  }

  // MARK: - Expo Lifecycle

  /// Called when JS mounts a listener for a specific event.
  func startObserving(eventName: String) {
    let queueCount = eventQueues[eventName]?.count ?? 0
    Log.call.debug(
      "Start observing - event: \(eventName), queuedEvents: \(queueCount)"
    )
    observingEvents.insert(eventName)
    flushQueue(for: eventName)
  }

  /// Called when JS unmounts the listener for a specific event.
  func stopObserving(eventName: String) {
    Log.call.debug("Stop observing - event: \(eventName)")
    observingEvents.remove(eventName)
  }

  // MARK: - Private

  private func buildEventBody(
    _ body: [String: Any],
    flushed: Bool,
    timestamp: Date
  ) -> [String: Any] {
    var result = body
    result["meta"] = [
      "flushed": flushed,
      "timestamp": ISO8601DateFormatter().string(from: timestamp),
    ]
    return result
  }

  private func queueEvent(name: String, body: [String: Any], timestamp: Date) {
    let limit = queueLimits[name] ?? defaultQueueLimit

    // If limit is 0, don't queue at all
    if limit == 0 {
      Log.call.debug("Dropping event (queueing disabled) - name: \(name)")
      return
    }

    var queue = eventQueues[name] ?? []
    queue.append(QueuedEvent(body: body, timestamp: timestamp))

    // Enforce limit by dropping oldest events if necessary
    if let limit = limit, queue.count > limit {
      let dropCount = queue.count - limit
      queue.removeFirst(dropCount)
      Log.call.debug(
        "Queueing event (dropped \(dropCount) old) - name: \(name), queueSize: \(queue.count)"
      )
    } else {
      Log.call.debug(
        "Queueing event (JS not listening) - name: \(name), queueSize: \(queue.count)"
      )
    }

    eventQueues[name] = queue
  }

  private func flushQueue(for eventName: String) {
    guard let queue = eventQueues[eventName], !queue.isEmpty,
      let module = module
    else {
      return
    }

    Log.call.debug(
      "Flushing event queue - event: \(eventName), count: \(queue.count)"
    )
    for event in queue {
      let body = buildEventBody(
        event.body,
        flushed: true,
        timestamp: event.timestamp
      )
      module.sendEvent(eventName, body)
    }

    eventQueues.removeValue(forKey: eventName)
  }
}
